projects/plutus/test/test_bps_heartbeat.c

```diff
@@ -0,0 +1,124 @@
+#include "bps_heartbeat.h"
+#include "can.h"
+#include "can_unpack.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "gpio.h"
+#include "interrupt.h"
+#include "log.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+#include "test_helpers.h"
+
+#define TEST_BPS_HEARTBEAT_PERIOD_MS 50
+#define TEST_BPS_HEARTBEAT_NUM_CAN_RX_HANDLERS 5
+
+static CANStorage s_can;
+static CANRxHandler s_rx_handlers[TEST_BPS_HEARTBEAT_NUM_CAN_RX_HANDLERS];
+static BpsHeartbeatStorage s_bps_heartbeat;
+static EERelayState s_relay_state;
+static CANAckStatus s_ack_status;
+static EEBpsHeartbeatState s_heartbeat_state;
+
+StatusCode TEST_MOCK(sequenced_relay_set_state)(SequencedRelayStorage *storage,
+                                                EERelayState state) {
+  LOG_DEBUG("Relay state: %d\n", state);
+  s_relay_state = state;
+
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_bps_rx(const CANMessage *msg, void *context, CANAckStatus *ack_reply) {
+  *ack_reply = s_ack_status;
+
+  uint8_t state = NUM_EE_BPS_HEARTBEAT_STATES;
+  CAN_UNPACK_BPS_HEARTBEAT(msg, &state);
+  LOG_DEBUG("ACK request: BPS state %d\n", state);
+  s_heartbeat_state = state;
+
+  return STATUS_CODE_OK;
+}
+
+typedef enum {
+  TEST_BPS_HEARTBEAT_EVENT_CAN_TX = 0,
+  TEST_BPS_HEARTBEAT_EVENT_CAN_RX,
+  TEST_BPS_HEARTBEAT_EVENT_CAN_FAULT,
+} TestBpsHeartbeatEvent;
+
+void setup_test(void) {
+  gpio_init();
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+
+  CANSettings settings = {
+    .device_id = SYSTEM_CAN_DEVICE_PLUTUS,
+    .bitrate = CAN_HW_BITRATE_500KBPS,
+    .rx_event = TEST_BPS_HEARTBEAT_EVENT_CAN_RX,
+    .tx_event = TEST_BPS_HEARTBEAT_EVENT_CAN_TX,
+    .fault_event = TEST_BPS_HEARTBEAT_EVENT_CAN_FAULT,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+    .loopback = true,
+  };
+
+  can_init(&settings, &s_can, s_rx_handlers, SIZEOF_ARRAY(s_rx_handlers));
+  TEST_ASSERT_OK(can_register_rx_handler(SYSTEM_CAN_MESSAGE_BPS_HEARTBEAT, prv_bps_rx, NULL));
+
+  // Closed relay by default
```

Comment:
For this test right? not for the car itself?

Reply:
Yeah, it's closed since we're testing that it opens when a fault occurs.

Reply:
Okay should probably note that as it might be confusing.